src/back/space/src/service/service.cpp

```diff
@@ -0,0 +1,327 @@
+}
+
+bool Service::IsLimitReached(const std::string& userId) {
+	return _spacesLimitForUser <= _repo.GetCountSpacesWithUser(userId);
+}
+
+void Service::Create(const std::string& name, const std::string& key, bool requestsAllowed, const std::string& userId) {
+	_repo.CreateSpaceAndItsOwner(name, key, requestsAllowed, userId);
+}
+
+void Service::Delete(const boost::uuids::uuid& id) {
+	_repo.SpaceUser().DeleteBySpace(id);
+	_repo.SpaceInvitation().DeleteBySpace(id);
+	_repo.SpaceLink().DeleteBySpace(id);
+	_repo.Space().Delete(id);
+}
+
+bool Service::IsSpaceOwner(const boost::uuids::uuid& id, const std::string& userId) {
+	return _repo.SpaceUser().IsOwner(id, userId);
+}
+
+void Service::Invite(const std::string& creatorId, const boost::uuids::uuid& spaceId, const std::string& userId, const Role::Type& role) {
+	_repo.Insert(spaceId, userId, role, creatorId);
+}
+
+void Service::ChangeRoleInInvitation(int id, const Role::Type& role, const std::string& userId) {
+	const auto invitation = _repo.SpaceInvitation().SelectById(id);
+	if (!_repo.SpaceUser().IsAdmin(invitation.spaceId, userId))
+		throw errors::Unauthorized401();
+
+	_repo.SpaceInvitation().UpdateRole(id, role);
+}
+
+void Service::ApproveInvitation(int id, const std::string& headerUserId) {
+	model::SpaceInvitation invitation = _repo.SpaceInvitation().SelectById(id);
+
+	static const std::set<Role::Type> valid_roles{
+		Role::User, Role::Guest, Role::Admin
+	};
+	if (!valid_roles.contains(invitation.role))
+		throw errors::BadRequest400("Wrong role");
+
+	// Я прошусь/хочет к нам - creatorId == userId
+	if (invitation.creatorId == invitation.userId) {
+		if (invitation.userId == headerUserId)
+			throw errors::Forbidden403();
+		// Я прошусь/хочет к нам - может одобрить только админ
+		if (!_repo.SpaceUser().IsAdmin(invitation.spaceId, headerUserId))
+			throw errors::Forbidden403();
+	}
+	// Меня/мы пригласили - может одобрить только пользователь которого пригласили
+	else if (invitation.userId != headerUserId)
+		throw errors::Forbidden403();
+
+	_repo.SpaceInvitation().DeleteById(id);
+
+	_repo.SpaceUser().Insert(invitation.spaceId, invitation.userId, false, invitation.role);
+}
+
+void Service::DeleteInvitation(int id, const std::string& headerUserId) {
+	const auto invitation = _repo.SpaceInvitation().SelectById(id);
+
+	bool isEnoughRights = false;
+
+	if (invitation.creatorId == invitation.userId && invitation.userId == headerUserId) {
+		// I want to join
+		isEnoughRights = true;
+	} else if (_repo.SpaceUser().IsAdmin(invitation.spaceId, headerUserId)) {
+		// other cases need admin rights
+		isEnoughRights = true;
+	}
+
+	if (!isEnoughRights)
+		throw errors::Unauthorized401();
+
+	_repo.SpaceInvitation().DeleteById(id);
+}
+
+bool Service::CheckExpiredAtValidity(std::chrono::system_clock::time_point expiredAt) {
+	// todo - is it right way to compare timestamps in this current situation?
+	const auto now = std::chrono::system_clock::now();
+	return (expiredAt > now);
+}
```

Comment:
Давай пожалуйста добавим const и ссылку аргументу, удалим коммент, и уберём скобки.

```suggestion
bool Service::CheckExpiredAtValidity(const std::chrono::system_clock::time_point& expiredAt) {
	return expiredAt > std::chrono::system_clock::now();
}
```

Reply:
Готово (думал, что std::chrono::system_clock::time_point - это 8 байт и ссылка не нужна).

Reply:
Да, ты прав. Тут уже давай оставим так.